### DISCUSSION TITLE
How does `vmIndex` work in src/nvidia/src/kernel/diagnostics/gpu_acct.c

### DISCUSSION THREAD

**Discussion by ccs100203** [NVIDIA/open-gpu-kernel-modules#313] [Q&A] [answered]
In function `gpuacctStopGpuAccounting_IMPL`, I see a variable `vmIndex` and it is assigned `NV_INVALID_VM_INDEX` at line 722. 
However, it will be checked next line immediately, so it seemed useless.
```
vmIndex = NV_INVALID_VM_INDEX;

if (vmIndex == NV_INVALID_VM_INDEX)
{
...
```
Moreover, I can't find any code about revising `vmIndex` except initialization. I'm curious how this variable and these functions like  `gpuacctStopGpuAccounting_IMPL` work.

Thanks.

**Comment by YusufKhan-gamedev** [comment]
Im guessing there was a use on the proprietary driver, but since all of that is specific to VMs which ogk does not support it was cut.

You can also see here:
`     pEntry->isGuestProcess = (vmIndex == NV_INVALID_VM_INDEX) ? NV_FALSE : NV_TRUE;
`
This seems to be sabotaging the use of VMs here. vmIndex may have been a variable used to check if a pid was a vm.

**Reply by ccs100203** [reply]
Thanks for your explanation.

**Comment by mtijanic** [comment, marked as answer]
Hi @ccs100203 , per the [driver readme](https://us.download.nvidia.com/XFree86/Linux-x86_64/515.57/README/kernel_open.html), vGPU is not supported from the open driver yet. @YusufKhan-gamedev is right that the same code is in use on the proprietary driver. In fact, much of the same "GPU Accounting" logic is used to measure utilization by a given VM and overall system utilization between CPU and GSP.

However, since the overall support for VGPU is missing, the code that determines the `vmIndex` was broken (and unreachable), so it was easiest to just remove it for the initial release. When vGPU support is brought up on the GSP driver model, I expect the missing code (or an alternative) to make an appearance in these files.

As a sidenote, this is also why you might see random `{}` blocks [like this](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/src/nvidia/src/kernel/diagnostics/gpu_acct.c#L423-L427) in this file. This might have been an `if (vgpu) {...} else {...}`, but the `if` block and `else` keyword were removed as the code they are referencing is not available. And related, this is (partially) why we don't accept cosmetic PRs currently, as someone would rightly want to remove the extra block, but that would make it incompatible with our p4 codebase.

**Reply by ccs100203** [reply]
Thanks for your reply. Although I feel a little frustrated in this situation, I still look forward to the future when Nvidia releases other drivers.
